_posts/2018-01-09-ble-virtualization.md

```diff
@@ -0,0 +1,125 @@
+
+  return 0;
+}
+
+```
+
+While running a Bluetooth Sniffer we find the four different Tock devices
+with different configurations as the screenshot below shows:
+
+
+![Bluetooth Low Energy](../assets/2018/01/ble_advertising.png)
+
+
+Note, that the kernel manages to generate unique advertisement addresses known
+as AdvA in the Bluetooth standard. Also, you can see that the applications by
+themselves configures different advertisement intervals and device names.
+TockOS2 is actually configured as Thermometer service with dummy data.
+
+
+## Implementation details
+First of all, we have not blogged about how the Bluetooth Low Energy driver works.
+The functionality is limited and support only advertisements and passive scanning so
+far i.e., no connections, different security modes and similar.
+
+
+Overall, the driver works in the following sequence:
+1. Initialize different Bluetooth parameters
+(advertisement interval, device name, other gap data and etc)
+2. Start advertise or start scanning, (this implies to configure virtual timer alarms)
+3. The virtual alarm expires and the radio gets activated for short while (magnitude in terms of milliseconds), for example by sending advertisements
+4. Reconfigure virtual alarm and go to sleep
+5. Step 3 and 4 continues forever until an application explicitly stops the operation
+
+
+However, in this blog post we want explain how we can use Tock architecture to
+achieve high degree of concurrency in the Bluetooth Low Energy driver and
+the different types of trade-offs we have performed with the implementation.
+A fundamental design decision of Tock is that the kernel is statically allocated
+which implies that the kernel can't dynamically allocate. Instead Tock uses an
+unique process model with five segments which the figure below shows:
+
+![Tock architecture](../assets/img/architecture.png)
+
+In context of this driver the grant segment is the essential part which makes
+it possible for the kernel to allocate memory and store data for each process.
+The grant contains Bluetooth Low Energy related data such as transmitting
+power, advertising interval, the actual data to advertise, and state along with
+virtual timer alarm.
+
+
+First, let's discuss the virtual alarm because it essential
+for correctness that the correct process is identified. Because, the virtual
+alarm doesn't indicate which process that has triggered the actual alarm requires
+iteration through the grants to find the expired process. Also, note that several
+alarms can expire at the same time with overlapping advertisement intervals or
+that the radio is busy once an alarm expires. In such scenarios, pseudo-randomness
+is introduced to to break ties and re-schedule later.   
+
+
+Second, the Bluetooth Low Energy related data stored in the grant must be written
+underlying chip before activating the chip because the order is before the
+expired alarm is not validated. This means, for example the actual radio buffer
```

Comment:
I'm not quite sure I follow what "because the order is before the expired alarm is not validated" is trying to say; try re-phrasing?

Reply:
Ahh, that was a poor explanation. I'm trying to say that the we don't know what the content is in the radio buffer so instead we replace it with the App's own buffer before starting advertising or scanning!